### ISSUE TITLE
Questions about AudioDevice behaviour

### ISSUE THREAD

**Issue by tejas-n** [twilio/video-quickstart-android#637]
### Description

We were following [this](https://github.com/twilio/video-quickstart-android/blob/master/exampleCustomAudioDevice/src/main/java/com/twilio/video/examples/examplecustomaudiodevice/FileAndMicAudioDevice.kt) sample to implement a custom audio device. We found that the audio started crackling whenever more than one remote participant joined the room. From further investigation we found out that `onStartRendering()` was called whenever a new remote participant joined the room. So a new thread was launched every time for rendering the audio to speakers .

https://github.com/twilio/video-quickstart-android/blob/62918484f096a1e4c6d7abf81877b147957468d5/exampleCustomAudioDevice/src/main/java/com/twilio/video/examples/examplecustomaudiodevice/FileAndMicAudioDevice.kt#L238-L247

Ignoring the subsequent calls to `onStartRendering()` using a flag caused the speaker to not play anything. What did work for us was stopping the rendering and reinitialising it explicitly on subsequent calls. We modified the code to  this and it works fine with multiple participants now:

```
   override fun onStartRendering(audioDeviceContext: AudioDeviceContext): Boolean {
        renderingAudioDeviceContext = audioDeviceContext
        // We added the following if block
        if (isRendering) {
            onStopRendering()
            onInitRenderer()
        }
        isRendering = true
        rendererThread = HandlerThread("RendererThread")
        rendererThread.start()
        // Create the capturer handler that processes the renderer Runnables.
        rendererHandler = Handler(rendererThread.looper)
        rendererHandler.post(speakerRendererRunnable)
        return true
    }
    
     override fun onStopRendering(): Boolean {
        isRendering = false
        stopAudioTrack()
        // Quit the rendererThread's looper to stop processing any further messages.
        rendererThread.quit()
        /*
         * When onStopRendering is called, the AudioDevice API expects that at the completion
         * of the callback the renderer has completely stopped. As a result, quit the renderer
         * thread and explicitly wait for the thread to complete.
         */
        if (!ThreadUtils.joinUninterruptibly(rendererThread, THREAD_JOIN_TIMEOUT_MS)) {
            Timber.w("Join of rendererThread timed out")
            return false
        }
        return true
    }

```

Is this the right approach to handle multiple calls to `onStartRendering()`? Can you tell us when exactly the `AudioDevice` methods are called by the SDK, please?

#### Video Android SDK

5.14.0

#### Android API

25

#### Android Device

A custom Android hardware

**Comment by tejas-n** [comment]
We slightly changed our approach. We found that even `onInitRenderer()` was being called(prior to `onStartRendering()`) for every participant that joined the room. So we changed the code to this and it works:

```
override fun onInitRenderer(): Boolean {
        // we added this if block
        if (isRendering) {
            onStopRendering()
        }
        val bytesPerFrame = rendererFormat.channelCount * (BITS_PER_SAMPLE / 8)
        readByteBuffer = ByteBuffer.allocateDirect(bytesPerFrame * (rendererFormat.sampleRate / BUFFERS_PER_SECOND))
        val channelConfig = channelCountToConfiguration(rendererFormat.channelCount)
        val minBufferSize = AudioRecord.getMinBufferSize(rendererFormat.sampleRate, channelConfig, android.media.AudioFormat.ENCODING_PCM_16BIT)
        audioTrack = AudioTrack(AudioManager.STREAM_VOICE_CALL, rendererFormat.sampleRate, channelConfig, android.media.AudioFormat.ENCODING_PCM_16BIT, minBufferSize, AudioTrack.MODE_STREAM)
        keepAliveRendererRunnable = true
        return true
    }
    
    override fun onStartRendering(audioDeviceContext: AudioDeviceContext): Boolean {
        renderingAudioDeviceContext = audioDeviceContext
        // Create the renderer thread and start
        //set isRendering to true
        isRendering = true
        rendererThread = HandlerThread("RendererThread")
        rendererThread.start()
        // Create the capturer handler that processes the renderer Runnables.
        rendererHandler = Handler(rendererThread.looper)
        rendererHandler.post(speakerRendererRunnable)
        return true
    }

    override fun onStopRendering(): Boolean {
       // set isRendering to false
        isRendering = false
        stopAudioTrack()
        // Quit the rendererThread's looper to stop processing any further messages.
        rendererThread.quit()
        /*
         * When onStopRendering is called, the AudioDevice API expects that at the completion
         * of the callback the renderer has completely stopped. As a result, quit the renderer
         * thread and explicitly wait for the thread to complete.
         */
        if (!ThreadUtils.joinUninterruptibly(rendererThread, THREAD_JOIN_TIMEOUT_MS)) {
            Timber.w("Join of rendererThread timed out")
            return false
        }
        return true
    }
```

**Comment by Alton09** [comment]
Thanks for reaching out @tejas-n ! Glad to see you found a solution. I'll close this one for now, but feel free to reopen as needed.

**Comment by jahead** [comment]
hey @Alton09 maybe this could be re-opened? and some documentation around this could be made? As we also spent some time trying to figure out this same issue. 

We also found that if you return false from `onInitRenderer() ` it still invokes `onStartRendering()`. It is unclear what the contracts are here from `AudioDeviceRenderer` interface. like what point is there to returning false, if your just going to invoke start... 

But our main issue here is that `onStopRenderer` isn't invoked before a init/start when more people join the call. 

What's more interesting is that the `WebRtcAudioTrack` from the defaultAudioDevice seems to make the same assumption. 

```
} else if (this.audioTrack != null) {
            this.reportWebRtcAudioTrackInitError("Conflict with existing AudioTrack.");
            return false;
} 
```

which is only cleaned up in the `WebRtcAudioTrack.stopPlayout`

 I get they aren't the same classes but it would be good to have some documents around when these callbacks are invoked, specifically if the sequence is different under different conditions.
